contracts/savings-app/src/handlers/execute.rs

```diff
@@ -297,3 +323,85 @@ pub(crate) fn _create_position(
         SubMsg::reply_always(create_msg, CREATE_POSITION_ID),
     ))
 }
+
+/// Sends autocompound rewards to the executor.
+/// In case user have not enough rewards it will also do a swap
+pub fn autocompound_executor_rewards(
+    deps: Deps,
+    env: &Env,
+    executor: String,
+    app: &App,
+    config: Config,
+) -> AppResult<Vec<CosmosMsg>> {
+    let rewards_config = config.autocompound_rewards_config;
+    let position = get_position(deps)?;
+    let user = position.owner;
+
+    // Get user balance of gas denom
+    let user_gas_balance = deps
+        .querier
+        .query_balance(user.clone(), rewards_config.gas_denom.clone())?;
+
+    let mut rewards_messages = vec![];
+
+    // If not enough gas coins - swap for some amount
+    if user_gas_balance.amount < rewards_config.min_gas_balance {
+        // Get asset entries
+        let dex = app.dex(deps, config.exchange);
+        let ans_host = app.ans_host(deps)?;
+        let gas_asset = AssetInfo::Native(rewards_config.gas_denom.clone())
+            .resolve(&deps.querier, &ans_host)?;
+        let swap_asset = AssetInfo::Native(rewards_config.swap_denom.clone())
+            .resolve(&deps.querier, &ans_host)?;
+
+        // Do reverse swap to find approximate amount we need to swap
+        let need_gas_coins = rewards_config.max_gas_balance - user_gas_balance.amount;
+        let simulate_swap_response = dex.simulate_swap(
+            AnsAsset::new(gas_asset.clone(), need_gas_coins),
+            swap_asset.clone(),
+        )?;
+
+        // Get user balance of swap denom
+        let user_swap_balance = deps
+            .querier
+            .query_balance(user.clone(), rewards_config.swap_denom)?;
+
+        // Swap as much as available if not enough for max_gas_balance
+        let swap_amount = simulate_swap_response
```

Comment:
In the case where there aren't enough gas tokens AND there aren't enough tokens to swap we should set some state in the contract that it's "pauzed" or something so we can prevent us from trying to execute it. Or do you think that the rewards query on the contract should figure that out without adding extra state?